### PR TITLE
Update from Dagger2.x to Android-Hilt.

### PR DESCRIPTION
**Pull request by ahmedeltaher**
Please review and give OK to merge

### CODE REVIEW

**Review by CodingRock-Star — APPROVED** [review]

**Review by PO15ON — DISMISSED** [review]
The PR seems good to me. No changes needed.

File: app/src/main/java/com/task/ui/component/details/DetailsViewModel.kt [hunk 1]
```diff
@@ -9,14 +9,15 @@ import com.task.data.Resource
...
 import kotlinx.coroutines.flow.collect
 import kotlinx.coroutines.launch
 import javax.inject.Inject
 
 /**
  * Created by AhmedEltaher
  */
-
+@HiltViewModel
 open class DetailsViewModel @Inject constructor(private val dataRepository: DataRepositorySource) : BaseViewModel() {
```
Comment (m04az) [inline app/src/main/java/com/task/ui/component/details/DetailsViewModel.kt:21]: https://developer.android.com/reference/androidx/hilt/lifecycle/ViewModelInject

Comment (ahmedeltaher) [reply app/src/main/java/com/task/ui/component/details/DetailsViewModel.kt:21]: Did you read the link you posted?!

Comment (m04az) [reply app/src/main/java/com/task/ui/component/details/DetailsViewModel.kt:21]: it is deprecated already :joy:
sorry man my mistake

**Review by Sabboo — APPROVED** [review]
Hilt migration is okay to me. 
However you should consider another update (isn't related to this PR) that will fix an existing issue that causes unit tests to fail as the provided json files cannot be found.

File: app/src/main/java/com/task/App.kt [hunk 2]
```diff
@@ -1,32 +1,18 @@
...
-    @Inject
-    lateinit var androidInjector: DispatchingAndroidInjector<Any>
-
-    override fun androidInjector(): AndroidInjector<Any> = androidInjector
+@HiltAndroidApp
+open class App : Application() {
 
     override fun onCreate() {
         super.onCreate()
         context = applicationContext
```
Comment (mrabelwahed) [inline app/src/main/java/com/task/App.kt:15]: I think you do not need to assign applicationContext to context variable here if you need it  activity for example you can make extension function, for Hilt it is all perfect Good Job Man

File: app/src/main/java/com/task/ui/component/login/LoginActivity.kt [hunk 3]
```diff
@@ -2,27 +2,24 @@ package com.task.ui.component.login
...
  */
-
+@AndroidEntryPoint
 class LoginActivity : BaseActivity() {
-    @Inject
-    lateinit var viewModelFactory: ViewModelFactory
 
-    @Inject
-    lateinit var loginViewModel: LoginViewModel
+    private val loginViewModel: LoginViewModel by viewModels()
```
Comment (Yasserkhayyal) [inline app/src/main/java/com/task/ui/component/login/LoginActivity.kt:22]: How did you manage to inject the viewModel with the default ViewModelFactory using Hilt given that It has "DataRepositorySource" as a dependency?! I can't see you using AbstractSavedStateViewModelFactory(owner, defaultArgs) as a factory for your viewModels

File: app/src/main/java/com/task/App.kt (hunk 2 above)
Comment (ahmedeltaher) [reply app/src/main/java/com/task/App.kt:15]: You are right ✔️

File: app/src/main/java/com/task/ui/component/login/LoginActivity.kt (hunk 3 above)
Comment (ahmedeltaher) [reply app/src/main/java/com/task/ui/component/login/LoginActivity.kt:22]: `by viewModels() ` Kotlin property delegate,  does that for you for free

**Review by mtzhisham — DISMISSED** [review]
General notes: 
- `App.context` could be replaced with `@ApplicationContext` `Context`
- For readability multiple interface/Impl  occurrence does not follow the one  naming pattern for example: 
  - `provideErrorFactoryImpl` and `ErrorFactory`
  - `provideErrorMapper` and `ErrorMapperInterface`

File: app/build.gradle [hunk 4]
```diff
@@ -142,8 +142,6 @@ dependencies {
     //MultiDex
     implementation "androidx.multidex:multidex:$multiDexVersion"
```
Comment (AmrIsmaiel) [inline app/build.gradle:143]: No Need for this library, you already is using min api 21

File: app/build.gradle [hunk 5]
```diff
@@ -3,10 +3,12 @@ apply plugin: 'kotlin-android'
 apply plugin: 'kotlin-android-extensions'
```
Comment (AmrIsmaiel) [inline app/build.gradle:3]: you are using DataBinding and applied kotlin-android-extensions plugin !!

File: build.gradle [hunk 6]
```diff
@@ -37,46 +37,46 @@ ext {
     // Sdk and tools
     // Support library and architecture components support minSdk 14 and above.
     minSdkVersion = 21
-    targetSdkVersion = 29
-    compileSdkVersion = 29
+    targetSdkVersion = 30
```
Comment (AmrIsmaiel) [inline build.gradle:40]: IMO, important ext needs to be in build.gradle(:app) for more and easy readability.

**Review by AmrIsmaiel — APPROVED** [review]
I think there are some minor changes but in general all is good.

File: build.gradle [hunk 7]
```diff
@@ -1,6 +1,6 @@
 // Top-level build file where you can add configuration options common to all sub-projects/modules.
 buildscript {
-    ext.kotlin_version = '1.4.10'
+    ext.kotlin_version = '1.4.21'
```
Comment (AhmedMousa7) [inline build.gradle:3]: I would recommend to take a look at this link before upgrading to version 1.4.21 of kotlin.

https://kotlinlang.org/docs/reference/whatsnew1420.html#kotlin-android-extensions

File: build.gradle [hunk 8]
```diff
@@ -37,46 +37,46 @@ ext {
...
     // Support library and architecture components support minSdk 14 and above.
     minSdkVersion = 21
-    targetSdkVersion = 29
-    compileSdkVersion = 29
+    targetSdkVersion = 30
+    compileSdkVersion = 30
 
-    junitVersion = '5.6.2'
-    junitRunnerVersion = '1.6.2'
+    junitVersion = '5.7.0'
```
[the remark this answers is not in the document]
Comment (ahmedeltaher) [reply build.gradle:43]: It is better to have it in sheared file between all your Application Modules
please take a look in google android arch sample

File: build.gradle (hunk 6 above)
Comment (ahmedeltaher) [reply build.gradle:40]: Take a look please in google arch samples and code labs

Comment (AmrIsmaiel) [reply build.gradle:40]: yes, I know they are doing the same you done so I wrote my opinion.
I just don't like switching between 2 files to see the versions specially `minSdk` and `targetSdk` :sweat_smile:

**Pull request merged**